solr/modules/llm/src/java/org/apache/solr/llm/embedding/SolrEmbeddingModel.java

```diff
@@ -0,0 +1,152 @@
+              break;
+            case MAX_SEGMENTS_PER_BATCH_PARAM:
+              builder
+                  .getClass()
+                  .getMethod(paramName, Integer.class)
+                  .invoke(builder, ((Long) params.get(paramName)).intValue());
+              break;
+            case MAX_RETRIES_PARAM:
+              builder
+                  .getClass()
+                  .getMethod(paramName, Integer.class)
+                  .invoke(builder, ((Long) params.get(paramName)).intValue());
+              break;
+            default:
+              ArrayList<Method> methods = new ArrayList<>();
+              for (var method : builder.getClass().getMethods()) {
+                if (paramName.equals(method.getName()) && method.getParameterCount() == 1) {
+                  methods.add(method);
+                }
+              }
+              if (methods.size() == 1) {
+                methods.get(0).invoke(builder, params.get(paramName));
+              } else {
+                builder
+                    .getClass()
+                    .getMethod(paramName, String.class)
+                    .invoke(builder, params.get(paramName));
+              }
+          }
+        }
+      }
+      embedder = (EmbeddingModel) builder.getClass().getMethod("build").invoke(builder);
+      return new SolrEmbeddingModel(name, embedder, params);
+    } catch (final Exception e) {
+      throw new EmbeddingModelException("Model loading failed for " + className, e);
+    }
+  }
+
+  public SolrEmbeddingModel(String name, EmbeddingModel embedder, Map<String, Object> params) {
+    this.name = name;
+    this.embedder = embedder;
+    this.params = params;
+    this.hashCode = calculateHashCode();
+  }
+
+  public float[] vectorise(String text) {
+    Embedding vector = embedder.embed(text).content();
+    return vector.vector();
+  }
+
+  @Override
+  public String toString() {
+    return getClass().getSimpleName() + "(name=" + getName() + ")";
+  }
+
+  @Override
+  public long ramBytesUsed() {
+    return BASE_RAM_BYTES
+        + RamUsageEstimator.sizeOfObject(name)
+        + RamUsageEstimator.sizeOfObject(embedder);
+  }
+
+  @Override
+  public int hashCode() {
+    if (hashCode == null) {
+      hashCode = calculateHashCode();
+    }
+    return hashCode;
+  }
+
+  private int calculateHashCode() {
+    final int prime = 31;
+    int result = 1;
+    result = (prime * result) + Objects.hashCode(name);
+    result = (prime * result) + Objects.hashCode(embedder);
+    return result;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (this == obj) return true;
+    if (!(obj instanceof SolrEmbeddingModel)) return false;
+    final SolrEmbeddingModel other = (SolrEmbeddingModel) obj;
+    return Objects.equals(embedder, other.embedder) && Objects.equals(name, other.name);
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public EmbeddingModel getEmbedder() {
+    return embedder;
+  }
```

Comment:
subjective: make only accessible what needs to be accessible
```suggestion
  public String getEmbedderClassName() {
    return embedder.getClass().getName();
  }
```